src/test/scala/com/lightbend/kafkalagexporter/LookupTableSpec.scala

```diff
@@ -4,16 +4,245 @@
+        table.lookup(301, redisClient) shouldBe Prediction(9000.3)
+      }
+
+      "lookups when table only contains a flat section with offsets same as lookup" in {
+        // Make sure the Point table is empty
+        redisClient.del(table.pointsKey)
+
+        table.addPoint(Point(0, 0), redisClient)
+        table.addPoint(Point(0, 100), redisClient)
+
+        table.lookup(0, redisClient) shouldBe LagIsZero
+      }
+
+      "lookup is zero when when table has a single element the same as the last group offset" in {
+        // Make sure the Point table is empty
+        redisClient.del(table.pointsKey)
+
+        table.addPoint(Point(0, 100), redisClient)
+        table.lookup(0, redisClient) shouldBe LagIsZero
+      }
+
+      "infinite lookups, dy == 0, flat curve/no growth" in {
+        // Make sure the Point table is empty
+        redisClient.del(table.pointsKey)
+
+        table.addPoint(Point(100, 100), redisClient)
+        table.addPoint(Point(100, 200), redisClient)
+        table.addPoint(Point(100, 300), redisClient)
+        table.addPoint(Point(100, 400), redisClient)
+
+        if (table.length(redisClient) != 2) {
+          fail(s"Expected flat entries to compress to a single entry $table")
+        }
+
+        if (table.mostRecentPoint(redisClient).right.get.time != 400) {
+          fail(s"Expected compressed table to have last timestamp $table")
+        }
+
+        table.lookup(99, redisClient) shouldBe Prediction(Double.NegativeInfinity)
+        table.lookup(101, redisClient) shouldBe Prediction(Double.PositiveInfinity)
+      }
+
+      "table retention and resolution" in {
+        val _config = ConsumerGroupCollector.CollectorConfig(0.second, 20, RedisConfig(enabled = true, resolution = 1.seconds, retention = 2.seconds, expiration = 30.minute), KafkaCluster("default", ""), Clock.fixed(Instant.ofEpochMilli(0), ZoneId.systemDefault()))
+        val table = Table(Domain.TopicPartition("topic", 0), _config).right.get
+
+        // Make sure the Point table is empty
+        redisClient.del(table.pointsKey)
+
+        table.addPoint(Point(100, Clock.systemUTC().instant().toEpochMilli), redisClient) shouldBe Inserted
+        table.addPoint(Point(200, Clock.systemUTC().instant().toEpochMilli), redisClient) shouldBe UpdatedRetention
+        Thread.sleep(1000)
+        table.addPoint(Point(200, Clock.systemUTC().instant().toEpochMilli), redisClient) shouldBe Inserted
+        Thread.sleep(1000)
+        table.addPoint(Point(200, Clock.systemUTC().instant().toEpochMilli), redisClient) shouldBe UpdatedSameOffset
+        table.addPoint(Point(300, Clock.systemUTC().instant().toEpochMilli), redisClient) shouldBe Inserted
+
+
+        if (table.length(redisClient) != 3) {
+          fail(s"Expected table to limit to 3 entries (current is ${table.length(redisClient)})")
+        }
+
+        // Sleeping 1 seconds for the first point to expire
+        Thread.sleep(1000)
```

Comment:
Please replace usages of `Thread.sleep` with ScalaTest `eventually` to block until the desired state is met. You can find a few examples throughout the tests.

https://www.scalatest.org/scaladoc/3.2.11/org/scalatest/concurrent/Eventually.html